Setting configs incompatible with Py_GIL_DISABLED should error harder

I noticed working on cryptography, which tries to build against the Python 3.7 limited API, that it has this pair of configs set:

```
cargo:rustc-cfg=Py_3_7
cargo:rustc-cfg=Py_GIL_DISABLED
```

This leads to confusing compiler errors due to the conditional compilation in pyo3-ffi not being written to handle this case since it's impossible:

```
DEBUG    Compiling cryptography-x509 v0.1.0 (/Users/goldbaum/Documents/cryptography/src/rust/cryptography-x509)
DEBUG error[E0432]: unresolved import `crate::PyMutex`
DEBUG  --> /Users/goldbaum/Documents/pyo3/pyo3-ffi/src/object.rs:3:5
DEBUG   |
DEBUG 3 | use crate::PyMutex;
DEBUG   |     ^^^^^^^^^^^^^^ no `PyMutex` in the root
DEBUG
DEBUG error[E0425]: cannot find value `_Py_IMMORTAL_REFCNT` in this scope
DEBUG    --> /Users/goldbaum/Documents/pyo3/pyo3-ffi/src/object.rs:162:20
DEBUG     |
DEBUG 29  | const _Py_IMMORTAL_REFCNT_LOCAL: u32 = u32::MAX;
DEBUG     | ------------------------------------------------ similarly named constant `_Py_IMMORTAL_REFCNT_LOCAL` defined here
DEBUG ...
DEBUG 162 |             return _Py_IMMORTAL_REFCNT;
DEBUG     |                    ^^^^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `_Py_IMMORTAL_REFCNT_LOCAL`
DEBUG
DEBUG error[E0560]: struct `object::PyObject` has no field named `ob_refcnt`
DEBUG   --> /Users/goldbaum/Documents/pyo3/pyo3-ffi/src/object.rs:55:5
DEBUG    |
DEBUG 55 |     ob_refcnt: 1,
DEBUG    |     ^^^^^^^^^ `object::PyObject` does not have this field
DEBUG    |
DEBUG    = note: all struct fields are already assigned
```

We should panic earlier because this configuration will never work.

--- a/guide/src/free-threading.md
+++ b/guide/src/free-threading.md
@@ -108,6 +108,15 @@ using single-phase initialization and the
 [`sequential`](https://github.com/PyO3/pyo3/tree/main/pyo3-ffi/examples/sequential)
 example for modules using multi-phase initialization.
 
+If you would like to use conditional compilation to trigger different code paths
+under the free-threaded build, you can use the `Py_GIL_DISABLED` attribute once
+you have configured your crate to generate the necessary build configuration
+data. See [the guide
+section](./building-and-distribution/multiple-python-versions.md) for more
+details about supporting multiple different Python versions, including the
+free-threaded build.
+
+
 ## Special considerations for the free-threaded build
 
 The free-threaded interpreter does not have a GIL, and this can make interacting
@@ -234,7 +243,24 @@ needed. For now you should explicitly add locking, possibly using conditional
 compilation or using the critical section API to avoid creating deadlocks with
 the GIL.
 
-## Thread-safe single initialization
+### Cannot build extensions using the limited API
+
+The free-threaded build uses a completely new ABI and there is not yet an
+equivalent to the limited API for the free-threaded ABI. That means if your
+crate depends on PyO3 using the `abi3` feature or an an `abi-pyxx` feature, PyO3
+will print a warning and ignore that setting when building extensions using the
+free-threaded interpreter.
+
+This means that if your package makes use of the ABI forward compatibility
+provided by the limited API to uploads only one wheel for each release of your
+package, you will need to update and tooling or instructions to also upload a
+version-specific free-threaded wheel.
+
+See [the guide section](./building-and-distribution/multiple-python-versions.md)
+for more details about supporting multiple different Python versions, including
+the free-threaded build.
+
+### Thread-safe single initialization
 
 Until version 0.23, PyO3 provided only [`GILOnceCell`] to enable deadlock-free
 single initialization of data in contexts that might execute arbitrary Python

--- a/guide/src/free-threading.md
+++ b/guide/src/free-threading.md
@@ -247,9 +247,9 @@ the GIL.
 
 The free-threaded build uses a completely new ABI and there is not yet an
 equivalent to the limited API for the free-threaded ABI. That means if your
-crate depends on PyO3 using the `abi3` feature or an an `abi-pyxx` feature, PyO3
-will print a warning and ignore that setting when building extensions using the
-free-threaded interpreter.
+crate depends on PyO3 using the `abi3` feature or an an `abi3-pyxx` feature,
+PyO3 will print a warning and ignore that setting when building extensions using
+the free-threaded interpreter.
 
 This means that if your package makes use of the ABI forward compatibility
 provided by the limited API to uploads only one wheel for each release of your

--- a/newsfragments/4719.fixed.md
+++ b/newsfragments/4719.fixed.md
@@ -1,1 +1,2 @@
-* Building extensions for the free-threaded ABI using Pythons ABIs older than 3.13 will now panic.
+* Fixed an issue that prevented building free-threaded extensions for crates
+  that request a specific minimum limited API version.

--- a/newsfragments/4719.fixed.md
+++ b/newsfragments/4719.fixed.md
@@ -1,0 +1,1 @@
+* Building extensions for the free-threaded ABI using Pythons ABIs older than 3.13 will now panic.

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -1650,9 +1650,6 @@ fn default_lib_name_windows(
         // https://github.com/python/cpython/issues/101614
         format!("python{}{}_d", version.major, version.minor)
     } else if abi3 && !(implementation.is_pypy() || implementation.is_graalpy()) {
-        if gil_disabled {
-            panic!("Free-threaded ABI does not support limited API extensions");
-        }
         if debug {
             WINDOWS_ABI3_DEBUG_LIB_NAME.to_owned()
         } else {
@@ -2511,35 +2508,6 @@ mod tests {
                 },
                 CPython,
                 false,
-                false,
-                false,
-                true,
-            )
-        })
-        .is_err());
-        // abi3 and free-threading are incompatible
-        assert!(std::panic::catch_unwind(|| {
-            super::default_lib_name_windows(
-                PythonVersion {
-                    major: 3,
-                    minor: 12,
-                },
-                CPython,
-                true,
-                false,
-                false,
-                true,
-            )
-        })
-        .is_err());
-        assert!(std::panic::catch_unwind(|| {
-            super::default_lib_name_windows(
-                PythonVersion {
-                    major: 3,
-                    minor: 13,
-                },
-                CPython,
-                true,
                 false,
                 false,
                 true,

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -691,6 +691,11 @@ pub struct PythonVersion {
 }
 
 impl PythonVersion {
+    pub const PY313: Self = PythonVersion {
+        major: 3,
+        minor: 13,
+    };
+    #[allow(dead_code)]
     const PY37: Self = PythonVersion { major: 3, minor: 7 };
 }
 
@@ -1606,11 +1611,9 @@ fn load_cross_compile_config(
     Ok(config)
 }
 
-// Link against python3.lib for the stable ABI on Windows.
-// See https://www.python.org/dev/peps/pep-0384/#linkage
-//
-// This contains only the limited ABI symbols.
+// These contains only the limited ABI symbols.
 const WINDOWS_ABI3_LIB_NAME: &str = "python3";
+const WINDOWS_ABI3_DEBUG_LIB_NAME: &str = "python3_d";
 
 fn default_lib_name_for_target(
     version: PythonVersion,
@@ -1642,16 +1645,19 @@ fn default_lib_name_windows(
     debug: bool,
     gil_disabled: bool,
 ) -> String {
-    if debug {
+    if debug && version.minor < 10 {
         // CPython bug: linking against python3_d.dll raises error
         // https://github.com/python/cpython/issues/101614
-        if gil_disabled {
-            format!("python{}{}t_d", version.major, version.minor)
-        } else {
-            format!("python{}{}_d", version.major, version.minor)
-        }
+        format!("python{}{}_d", version.major, version.minor)
     } else if abi3 && !(implementation.is_pypy() || implementation.is_graalpy()) {
-        WINDOWS_ABI3_LIB_NAME.to_owned()
+        if gil_disabled {
+            panic!("Free-threaded ABI does not support limited API extensions");
+        }
+        if debug {
+            WINDOWS_ABI3_DEBUG_LIB_NAME.to_owned()
+        } else {
+            WINDOWS_ABI3_LIB_NAME.to_owned()
+        }
     } else if mingw {
         if gil_disabled {
             panic!("MinGW free-threaded builds are not currently tested or supported")
@@ -1659,7 +1665,16 @@ fn default_lib_name_windows(
         // https://packages.msys2.org/base/mingw-w64-python
         format!("python{}.{}", version.major, version.minor)
     } else if gil_disabled {
-        format!("python{}{}t", version.major, version.minor)
+        if version.major <= 3 && version.minor < 13 {
+            panic!("Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
+        }
+        if debug {
+            format!("python{}{}t_d", version.major, version.minor)
+        } else {
+            format!("python{}{}t", version.major, version.minor)
+        }
+    } else if debug {
+        format!("python{}{}_d", version.major, version.minor)
     } else {
         format!("python{}{}", version.major, version.minor)
     }
@@ -1676,8 +1691,10 @@ fn default_lib_name_unix(
             Some(ld_version) => format!("python{}", ld_version),
             None => {
                 if version > PythonVersion::PY37 {
-                    // PEP 3149 ABI version tags are finally gone
                     if gil_disabled {
+                        if version < PythonVersion::PY313 {
+                            panic!("Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
+                        }
                         format!("python{}.{}t", version.major, version.minor)
                     } else {
                         format!("python{}.{}", version.major, version.minor)
@@ -2392,6 +2409,17 @@ mod tests {
             ),
             "python39",
         );
+        assert!(std::panic::catch_unwind(|| {
+            super::default_lib_name_windows(
+                PythonVersion { major: 3, minor: 9 },
+                CPython,
+                false,
+                false,
+                false,
+                true,
+            )
+        })
+        .is_err());
         assert_eq!(
             super::default_lib_name_windows(
                 PythonVersion { major: 3, minor: 9 },
@@ -2447,7 +2475,7 @@ mod tests {
             ),
             "python39_d",
         );
-        // abi3 debug builds on windows use version-specific lib
+        // abi3 debug builds on windows use version-specific lib on 3.9 and older
         // to workaround https://github.com/python/cpython/issues/101614
         assert_eq!(
             super::default_lib_name_windows(
@@ -2459,6 +2487,107 @@ mod tests {
                 false,
             ),
             "python39_d",
+        );
+        assert_eq!(
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 10
+                },
+                CPython,
+                true,
+                false,
+                true,
+                false,
+            ),
+            "python3_d",
+        );
+        // Python versions older than 3.13 panic if gil_disabled is true
+        assert!(std::panic::catch_unwind(|| {
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 12,
+                },
+                CPython,
+                false,
+                false,
+                false,
+                true,
+            )
+        })
+        .is_err());
+        // abi3 and free-threading are incompatible
+        assert!(std::panic::catch_unwind(|| {
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 12,
+                },
+                CPython,
+                true,
+                false,
+                false,
+                true,
+            )
+        })
+        .is_err());
+        assert!(std::panic::catch_unwind(|| {
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 13,
+                },
+                CPython,
+                true,
+                false,
+                false,
+                true,
+            )
+        })
+        .is_err());
+        // mingw and free-threading are incompatible (until someone adds support)
+        assert!(std::panic::catch_unwind(|| {
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 12,
+                },
+                CPython,
+                false,
+                true,
+                false,
+                true,
+            )
+        })
+        .is_err());
+        assert_eq!(
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 13
+                },
+                CPython,
+                false,
+                false,
+                false,
+                true,
+            ),
+            "python313t",
+        );
+        assert_eq!(
+            super::default_lib_name_windows(
+                PythonVersion {
+                    major: 3,
+                    minor: 13
+                },
+                CPython,
+                false,
+                false,
+                true,
+                true,
+            ),
+            "python313t_d",
         );
     }
 
@@ -2520,6 +2649,33 @@ mod tests {
             ),
             "pypy3.9d-c",
         );
+
+        // free-threading adds a t suffix
+        assert_eq!(
+            super::default_lib_name_unix(
+                PythonVersion {
+                    major: 3,
+                    minor: 13
+                },
+                CPython,
+                None,
+                true
+            ),
+            "python3.13t",
+        );
+        // 3.12 and older are incompatible with gil_disabled
+        assert!(std::panic::catch_unwind(|| {
+            super::default_lib_name_unix(
+                PythonVersion {
+                    major: 3,
+                    minor: 12,
+                },
+                CPython,
+                None,
+                true,
+            )
+        })
+        .is_err());
     }
 
     #[test]

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -660,10 +660,14 @@ print("gil_disabled", get_config_var("Py_GIL_DISABLED"))
         )
     }
 
-    /// Lowers the configured version to the abi3 version, if set.
+    /// Updates configured ABI to build for to the requested abi3 version
+    /// This is a no-op for platforms where abi3 is not supported
     fn fixup_for_abi3_version(&mut self, abi3_version: Option<PythonVersion>) -> Result<()> {
-        // PyPy doesn't support abi3; don't adjust the version
-        if self.implementation.is_pypy() || self.implementation.is_graalpy() {
+        // PyPy, GraalPy, and the free-threaded build don't support abi3; don't adjust the version
+        if self.implementation.is_pypy()
+            || self.implementation.is_graalpy()
+            || self.build_flags.0.contains(&BuildFlag::Py_GIL_DISABLED)
+        {
             return Ok(());
         }
 

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -176,7 +176,7 @@ impl InterpreterConfig {
         }
 
         // If Py_GIL_DISABLED is set, do not build with limited API support
-        if self.abi3 && !self.build_flags.0.contains(&BuildFlag::Py_GIL_DISABLED) {
+        if self.abi3 && !self.is_free_threaded() {
             out.push("cargo:rustc-cfg=Py_LIMITED_API".to_owned());
         }
 
@@ -660,13 +660,17 @@ print("gil_disabled", get_config_var("Py_GIL_DISABLED"))
         )
     }
 
+    pub fn is_free_threaded(&self) -> bool {
+        self.build_flags.0.contains(&BuildFlag::Py_GIL_DISABLED)
+    }
+
     /// Updates configured ABI to build for to the requested abi3 version
     /// This is a no-op for platforms where abi3 is not supported
     fn fixup_for_abi3_version(&mut self, abi3_version: Option<PythonVersion>) -> Result<()> {
         // PyPy, GraalPy, and the free-threaded build don't support abi3; don't adjust the version
         if self.implementation.is_pypy()
             || self.implementation.is_graalpy()
-            || self.build_flags.0.contains(&BuildFlag::Py_GIL_DISABLED)
+            || self.is_free_threaded()
         {
             return Ok(());
         }

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -1662,7 +1662,7 @@ fn default_lib_name_windows(
         // https://packages.msys2.org/base/mingw-w64-python
         format!("python{}.{}", version.major, version.minor)
     } else if gil_disabled {
-        if version.major <= 3 && version.minor < 13 {
+        if version < PythonVersion::PY313 {
             panic!("Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
         }
         if debug {

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -2507,7 +2507,7 @@ mod tests {
             .unwrap(),
             "python3_d",
         );
-        // Python versions older than 3.13 panic if gil_disabled is true
+        // Python versions older than 3.13 don't support gil_disabled
         assert!(super::default_lib_name_windows(
             PythonVersion {
                 major: 3,

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -309,14 +309,14 @@ print("gil_disabled", get_config_var("Py_GIL_DISABLED"))
                 // `_d.cp312-win_amd64.pyd` for 3.12 debug build
                 map["ext_suffix"].starts_with("_d."),
                 gil_disabled,
-            )
+            )?
         } else {
             default_lib_name_unix(
                 version,
                 implementation,
                 map.get("ld_version").map(String::as_str),
                 gil_disabled,
-            )
+            )?
         };
 
         let lib_dir = if cfg!(windows) {
@@ -394,7 +394,7 @@ print("gil_disabled", get_config_var("Py_GIL_DISABLED"))
             implementation,
             sysconfigdata.get_value("LDVERSION"),
             gil_disabled,
-        ));
+        )?);
         let pointer_width = parse_key!(sysconfigdata, "SIZEOF_VOID_P")
             .map(|bytes_width: u32| bytes_width * 8)
             .ok();
@@ -1544,7 +1544,7 @@ fn default_cross_compile(cross_compile_config: &CrossCompileConfig) -> Result<In
 /// Windows, macOS and Linux.
 ///
 /// Must be called from a PyO3 crate build script.
-fn default_abi3_config(host: &Triple, version: PythonVersion) -> InterpreterConfig {
+fn default_abi3_config(host: &Triple, version: PythonVersion) -> Result<InterpreterConfig> {
     // FIXME: PyPy & GraalPy do not support the Stable ABI.
     let implementation = PythonImplementation::CPython;
     let abi3 = true;
@@ -1557,12 +1557,12 @@ fn default_abi3_config(host: &Triple, version: PythonVersion) -> InterpreterConf
             false,
             false,
             false,
-        ))
+        )?)
     } else {
         None
     };
 
-    InterpreterConfig {
+    Ok(InterpreterConfig {
         implementation,
         version,
         shared: true,
@@ -1574,7 +1574,7 @@ fn default_abi3_config(host: &Triple, version: PythonVersion) -> InterpreterConf
         build_flags: BuildFlags::default(),
         suppress_build_script_link_lines: false,
         extra_build_script_lines: vec![],
-    }
+    })
 }
 
 /// Detects the cross compilation target interpreter configuration from all
@@ -1625,16 +1625,9 @@ fn default_lib_name_for_target(
     target: &Triple,
 ) -> Option<String> {
     if target.operating_system == OperatingSystem::Windows {
-        Some(default_lib_name_windows(
-            version,
-            implementation,
-            abi3,
-            false,
-            false,
-            false,
-        ))
+        Some(default_lib_name_windows(version, implementation, abi3, false, false, false).unwrap())
     } else if is_linking_libpython_for_target(target) {
-        Some(default_lib_name_unix(version, implementation, None, false))
+        Some(default_lib_name_unix(version, implementation, None, false).unwrap())
     } else {
         None
     }
@@ -1647,36 +1640,35 @@ fn default_lib_name_windows(
     mingw: bool,
     debug: bool,
     gil_disabled: bool,
-) -> String {
+) -> Result<String> {
     if debug && version < PythonVersion::PY310 {
         // CPython bug: linking against python3_d.dll raises error
         // https://github.com/python/cpython/issues/101614
-        format!("python{}{}_d", version.major, version.minor)
+        Ok(format!("python{}{}_d", version.major, version.minor))
     } else if abi3 && !(implementation.is_pypy() || implementation.is_graalpy()) {
         if debug {
-            WINDOWS_ABI3_DEBUG_LIB_NAME.to_owned()
+            Ok(WINDOWS_ABI3_DEBUG_LIB_NAME.to_owned())
         } else {
-            WINDOWS_ABI3_LIB_NAME.to_owned()
+            Ok(WINDOWS_ABI3_LIB_NAME.to_owned())
         }
     } else if mingw {
-        if gil_disabled {
-            panic!("MinGW free-threaded builds are not currently tested or supported")
-        }
+        ensure!(
+            !gil_disabled,
+            "MinGW free-threaded builds are not currently tested or supported"
+        );
         // https://packages.msys2.org/base/mingw-w64-python
-        format!("python{}.{}", version.major, version.minor)
+        Ok(format!("python{}.{}", version.major, version.minor))
     } else if gil_disabled {
-        if version < PythonVersion::PY313 {
-            panic!("Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
-        }
+        ensure!(version >= PythonVersion::PY313, "Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
         if debug {
-            format!("python{}{}t_d", version.major, version.minor)
+            Ok(format!("python{}{}t_d", version.major, version.minor))
         } else {
-            format!("python{}{}t", version.major, version.minor)
+            Ok(format!("python{}{}t", version.major, version.minor))
         }
     } else if debug {
-        format!("python{}{}_d", version.major, version.minor)
+        Ok(format!("python{}{}_d", version.major, version.minor))
     } else {
-        format!("python{}{}", version.major, version.minor)
+        Ok(format!("python{}{}", version.major, version.minor))
     }
 }
 
@@ -1685,33 +1677,31 @@ fn default_lib_name_unix(
     implementation: PythonImplementation,
     ld_version: Option<&str>,
     gil_disabled: bool,
-) -> String {
+) -> Result<String> {
     match implementation {
         PythonImplementation::CPython => match ld_version {
-            Some(ld_version) => format!("python{}", ld_version),
+            Some(ld_version) => Ok(format!("python{}", ld_version)),
             None => {
                 if version > PythonVersion::PY37 {
                     // PEP 3149 ABI version tags are finally gone
                     if gil_disabled {
-                        if version < PythonVersion::PY313 {
-                            panic!("Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
-                        }
-                        format!("python{}.{}t", version.major, version.minor)
+                        ensure!(version >= PythonVersion::PY313, "Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);
+                        Ok(format!("python{}.{}t", version.major, version.minor))
                     } else {
-                        format!("python{}.{}", version.major, version.minor)
+                        Ok(format!("python{}.{}", version.major, version.minor))
                     }
                 } else {
                     // Work around https://bugs.python.org/issue36707
-                    format!("python{}.{}m", version.major, version.minor)
+                    Ok(format!("python{}.{}m", version.major, version.minor))
                 }
             }
         },
         PythonImplementation::PyPy => match ld_version {
-            Some(ld_version) => format!("pypy{}-c", ld_version),
-            None => format!("pypy{}.{}-c", version.major, version.minor),
+            Some(ld_version) => Ok(format!("pypy{}-c", ld_version)),
+            None => Ok(format!("pypy{}.{}-c", version.major, version.minor)),
         },
 
-        PythonImplementation::GraalPy => "python-native".to_string(),
+        PythonImplementation::GraalPy => Ok("python-native".to_string()),
     }
 }
 
@@ -1881,7 +1871,7 @@ pub fn make_interpreter_config() -> Result<InterpreterConfig> {
         );
     };
 
-    let mut interpreter_config = default_abi3_config(&host, abi3_version.unwrap());
+    let mut interpreter_config = default_abi3_config(&host, abi3_version.unwrap())?;
 
     // Auto generate python3.dll import libraries for Windows targets.
     #[cfg(feature = "python3-dll-a")]
@@ -2218,7 +2208,7 @@ mod tests {
         let min_version = "3.7".parse().unwrap();
 
         assert_eq!(
-            default_abi3_config(&host, min_version),
+            default_abi3_config(&host, min_version).unwrap(),
             InterpreterConfig {
                 implementation: PythonImplementation::CPython,
                 version: PythonVersion { major: 3, minor: 7 },
@@ -2241,7 +2231,7 @@ mod tests {
         let min_version = "3.9".parse().unwrap();
 
         assert_eq!(
-            default_abi3_config(&host, min_version),
+            default_abi3_config(&host, min_version).unwrap(),
             InterpreterConfig {
                 implementation: PythonImplementation::CPython,
                 version: PythonVersion { major: 3, minor: 9 },
@@ -2407,19 +2397,18 @@ mod tests {
                 false,
                 false,
                 false,
-            ),
+            )
+            .unwrap(),
             "python39",
         );
-        assert!(std::panic::catch_unwind(|| {
-            super::default_lib_name_windows(
-                PythonVersion { major: 3, minor: 9 },
-                CPython,
-                false,
-                false,
-                false,
-                true,
-            )
-        })
+        assert!(super::default_lib_name_windows(
+            PythonVersion { major: 3, minor: 9 },
+            CPython,
+            false,
+            false,
+            false,
+            true,
+        )
         .is_err());
         assert_eq!(
             super::default_lib_name_windows(
@@ -2429,7 +2418,8 @@ mod tests {
                 false,
                 false,
                 false,
-            ),
+            )
+            .unwrap(),
             "python3",
         );
         assert_eq!(
@@ -2440,7 +2430,8 @@ mod tests {
                 true,
                 false,
                 false,
-            ),
+            )
+            .unwrap(),
             "python3.9",
         );
         assert_eq!(
@@ -2451,7 +2442,8 @@ mod tests {
                 true,
                 false,
                 false,
-            ),
+            )
+            .unwrap(),
             "python3",
         );
         assert_eq!(
@@ -2462,7 +2454,8 @@ mod tests {
                 false,
                 false,
                 false,
-            ),
+            )
+            .unwrap(),
             "python39",
         );
         assert_eq!(
@@ -2473,7 +2466,8 @@ mod tests {
                 false,
                 true,
                 false,
-            ),
+            )
+            .unwrap(),
             "python39_d",
         );
         // abi3 debug builds on windows use version-specific lib on 3.9 and older
@@ -2486,7 +2480,8 @@ mod tests {
                 false,
                 true,
                 false,
-            ),
+            )
+            .unwrap(),
             "python39_d",
         );
         assert_eq!(
@@ -2500,38 +2495,35 @@ mod tests {
                 false,
                 true,
                 false,
-            ),
+            )
+            .unwrap(),
             "python3_d",
         );
         // Python versions older than 3.13 panic if gil_disabled is true
-        assert!(std::panic::catch_unwind(|| {
-            super::default_lib_name_windows(
-                PythonVersion {
-                    major: 3,
-                    minor: 12,
-                },
-                CPython,
-                false,
-                false,
-                false,
-                true,
-            )
-        })
+        assert!(super::default_lib_name_windows(
+            PythonVersion {
+                major: 3,
+                minor: 12,
+            },
+            CPython,
+            false,
+            false,
+            false,
+            true,
+        )
         .is_err());
         // mingw and free-threading are incompatible (until someone adds support)
-        assert!(std::panic::catch_unwind(|| {
-            super::default_lib_name_windows(
-                PythonVersion {
-                    major: 3,
-                    minor: 12,
-                },
-                CPython,
-                false,
-                true,
-                false,
-                true,
-            )
-        })
+        assert!(super::default_lib_name_windows(
+            PythonVersion {
+                major: 3,
+                minor: 12,
+            },
+            CPython,
+            false,
+            true,
+            false,
+            true,
+        )
         .is_err());
         assert_eq!(
             super::default_lib_name_windows(
@@ -2544,7 +2536,8 @@ mod tests {
                 false,
                 false,
                 true,
-            ),
+            )
+            .unwrap(),
             "python313t",
         );
         assert_eq!(
@@ -2558,7 +2551,8 @@ mod tests {
                 false,
                 true,
                 true,
-            ),
+            )
+            .unwrap(),
             "python313t_d",
         );
     }
@@ -2573,7 +2567,8 @@ mod tests {
                 CPython,
                 None,
                 false
-            ),
+            )
+            .unwrap(),
             "python3.7m",
         );
         // Defaults to pythonX.Y for CPython 3.8+
@@ -2583,7 +2578,8 @@ mod tests {
                 CPython,
                 None,
                 false
-            ),
+            )
+            .unwrap(),
             "python3.8",
         );
         assert_eq!(
@@ -2592,7 +2588,8 @@ mod tests {
                 CPython,
                 None,
                 false
-            ),
+            )
+            .unwrap(),
             "python3.9",
         );
         // Can use ldversion to override for CPython
@@ -2602,13 +2599,15 @@ mod tests {
                 CPython,
                 Some("3.7md"),
                 false
-            ),
+            )
+            .unwrap(),
             "python3.7md",
         );
 
         // PyPy 3.9 includes ldversion
         assert_eq!(
-            super::default_lib_name_unix(PythonVersion { major: 3, minor: 9 }, PyPy, None, false),
+            super::default_lib_name_unix(PythonVersion { major: 3, minor: 9 }, PyPy, None, false)
+                .unwrap(),
             "pypy3.9-c",
         );
 
@@ -2618,7 +2617,8 @@ mod tests {
                 PyPy,
                 Some("3.9d"),
                 false
-            ),
+            )
+            .unwrap(),
             "pypy3.9d-c",
         );
 
@@ -2632,21 +2632,20 @@ mod tests {
                 CPython,
                 None,
                 true
-            ),
+            )
+            .unwrap(),
             "python3.13t",
         );
         // 3.12 and older are incompatible with gil_disabled
-        assert!(std::panic::catch_unwind(|| {
-            super::default_lib_name_unix(
-                PythonVersion {
-                    major: 3,
-                    minor: 12,
-                },
-                CPython,
-                None,
-                true,
-            )
-        })
+        assert!(super::default_lib_name_unix(
+            PythonVersion {
+                major: 3,
+                minor: 12,
+            },
+            CPython,
+            None,
+            true,
+        )
         .is_err());
     }
 

--- a/pyo3-build-config/src/impl_.rs
+++ b/pyo3-build-config/src/impl_.rs
@@ -695,7 +695,10 @@ impl PythonVersion {
         major: 3,
         minor: 13,
     };
-    #[allow(dead_code)]
+    const PY310: Self = PythonVersion {
+        major: 3,
+        minor: 10,
+    };
     const PY37: Self = PythonVersion { major: 3, minor: 7 };
 }
 
@@ -1645,7 +1648,7 @@ fn default_lib_name_windows(
     debug: bool,
     gil_disabled: bool,
 ) -> String {
-    if debug && version.minor < 10 {
+    if debug && version < PythonVersion::PY310 {
         // CPython bug: linking against python3_d.dll raises error
         // https://github.com/python/cpython/issues/101614
         format!("python{}{}_d", version.major, version.minor)
@@ -1688,6 +1691,7 @@ fn default_lib_name_unix(
             Some(ld_version) => format!("python{}", ld_version),
             None => {
                 if version > PythonVersion::PY37 {
+                    // PEP 3149 ABI version tags are finally gone
                     if gil_disabled {
                         if version < PythonVersion::PY313 {
                             panic!("Cannot compile C extensions for the free-threaded build on Python versions earlier than 3.13, found {}.{}", version.major, version.minor);

--- a/pyo3-ffi/build.rs
+++ b/pyo3-ffi/build.rs
@@ -113,11 +113,7 @@ fn ensure_python_version(interpreter_config: &InterpreterConfig) -> Result<()> {
     if interpreter_config.abi3 {
         match interpreter_config.implementation {
             PythonImplementation::CPython => {
-                if interpreter_config
-                    .build_flags
-                    .0
-                    .contains(&BuildFlag::Py_GIL_DISABLED)
-                {
+                if interpreter_config.is_free_threaded() {
                     warn!(
                             "The free-threaded build of CPython does not yet support abi3 so the build artifacts will be version-specific."
                     )

--- a/pyo3-ffi/build.rs
+++ b/pyo3-ffi/build.rs
@@ -118,16 +118,9 @@ fn ensure_python_version(interpreter_config: &InterpreterConfig) -> Result<()> {
                     .0
                     .contains(&BuildFlag::Py_GIL_DISABLED)
                 {
-                    let version = interpreter_config.version;
-                    if version < PythonVersion::PY313 {
-                        panic!(
-                            "The free-threaded ABI is not defined for versions older than 3.13, tried to build using Python {}.{} ABI", version.major, version.minor
-                        );
-                    } else {
-                        warn!(
+                    warn!(
                             "The free-threaded build of CPython does not yet support abi3 so the build artifacts will be version-specific."
-                        )
-                    }
+                    )
                 }
             }
             PythonImplementation::PyPy => warn!(

--- a/pyo3-ffi/build.rs
+++ b/pyo3-ffi/build.rs
@@ -1,11 +1,10 @@
 use pyo3_build_config::{
     bail, ensure, print_feature_cfgs,
     pyo3_build_script_impl::{
-        cargo_env_var, env_var,
-        errors::{Error, Result},
-        is_linking_libpython, resolve_interpreter_config, InterpreterConfig, PythonVersion,
+        cargo_env_var, env_var, errors::Result, is_linking_libpython, resolve_interpreter_config,
+        InterpreterConfig, PythonVersion,
     },
-    warn, BuildFlag, PythonImplementation,
+    warn, PythonImplementation,
 };
 
 /// Minimum Python version PyO3 supports.

--- a/pyo3-ffi/build.rs
+++ b/pyo3-ffi/build.rs
@@ -112,9 +112,16 @@ fn ensure_python_version(interpreter_config: &InterpreterConfig) -> Result<()> {
                     .0
                     .contains(&BuildFlag::Py_GIL_DISABLED)
                 {
-                    warn!(
+                    let version = interpreter_config.version;
+                    if version < PythonVersion::PY313 {
+                        panic!(
+                            "The free-threaded ABI is not defined for versions older than 3.13, tried to build using Python {}.{} ABI", version.major, version.minor
+                        );
+                    } else {
+                        warn!(
                             "The free-threaded build of CPython does not yet support abi3 so the build artifacts will be version-specific."
                         )
+                    }
                 }
             }
             PythonImplementation::PyPy => warn!(

--- a/pyo3-ffi/build.rs
+++ b/pyo3-ffi/build.rs
@@ -56,15 +56,21 @@ fn ensure_python_version(interpreter_config: &InterpreterConfig) -> Result<()> {
                 interpreter_config.version,
                 versions.min,
             );
-            ensure!(
-                interpreter_config.version <= versions.max || env_var("PYO3_USE_ABI3_FORWARD_COMPATIBILITY").map_or(false, |os_str| os_str == "1"),
-                "the configured Python interpreter version ({}) is newer than PyO3's maximum supported version ({})\n\
-                 = help: please check if an updated version of PyO3 is available. Current version: {}\n\
-                 = help: set PYO3_USE_ABI3_FORWARD_COMPATIBILITY=1 to suppress this check and build anyway using the stable ABI",
-                interpreter_config.version,
-                versions.max,
-                std::env::var("CARGO_PKG_VERSION").unwrap(),
-            );
+            if !interpreter_config
+                .build_flags
+                .0
+                .contains(&BuildFlag::Py_GIL_DISABLED)
+            {
+                ensure!(
+                        interpreter_config.version <= versions.max || env_var("PYO3_USE_ABI3_FORWARD_COMPATIBILITY").map_or(false, |os_str| os_str == "1"),
+                        "the configured Python interpreter version ({}) is newer than PyO3's maximum supported version ({})\n\
+                         = help: please check if an updated version of PyO3 is available. Current version: {}\n\
+                         = help: set PYO3_USE_ABI3_FORWARD_COMPATIBILITY=1 to suppress this check and build anyway using the stable ABI",
+                        interpreter_config.version,
+                        versions.max,
+                        std::env::var("CARGO_PKG_VERSION").unwrap(),
+                    );
+            }
         }
         PythonImplementation::PyPy => {
             let versions = SUPPORTED_VERSIONS_PYPY;

--- a/pyo3-ffi/build.rs
+++ b/pyo3-ffi/build.rs
@@ -1,8 +1,9 @@
 use pyo3_build_config::{
     bail, ensure, print_feature_cfgs,
     pyo3_build_script_impl::{
-        cargo_env_var, env_var, errors::Result, is_linking_libpython, resolve_interpreter_config,
-        InterpreterConfig, PythonVersion,
+        cargo_env_var, env_var,
+        errors::{Error, Result},
+        is_linking_libpython, resolve_interpreter_config, InterpreterConfig, PythonVersion,
     },
     warn, BuildFlag, PythonImplementation,
 };
@@ -56,20 +57,21 @@ fn ensure_python_version(interpreter_config: &InterpreterConfig) -> Result<()> {
                 interpreter_config.version,
                 versions.min,
             );
-            if !interpreter_config
-                .build_flags
-                .0
-                .contains(&BuildFlag::Py_GIL_DISABLED)
-            {
-                ensure!(
-                        interpreter_config.version <= versions.max || env_var("PYO3_USE_ABI3_FORWARD_COMPATIBILITY").map_or(false, |os_str| os_str == "1"),
+            if interpreter_config.version > versions.max {
+                ensure!(!interpreter_config.is_free_threaded(),
+                        "The configured Python interpreter version ({}) is newer than PyO3's maximum supported version ({})\n\
+                         = help: please check if an updated version of PyO3 is available. Current version: {}\n\
+                         = help: The free-threaded build of CPython does not support the limited API so this check cannot be suppressed.",
+                        interpreter_config.version, versions.max, std::env::var("CARGO_PKG_VERSION").unwrap()
+                );
+                ensure!(env_var("PYO3_USE_ABI3_FORWARD_COMPATIBILITY").map_or(false, |os_str| os_str == "1"),
                         "the configured Python interpreter version ({}) is newer than PyO3's maximum supported version ({})\n\
                          = help: please check if an updated version of PyO3 is available. Current version: {}\n\
                          = help: set PYO3_USE_ABI3_FORWARD_COMPATIBILITY=1 to suppress this check and build anyway using the stable ABI",
                         interpreter_config.version,
                         versions.max,
                         std::env::var("CARGO_PKG_VERSION").unwrap(),
-                    );
+                );
             }
         }
         PythonImplementation::PyPy => {

